sys/net/transport_layer/ng_udp/ng_udp.c

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     net_ng_udp
+ * @{
+ *
+ * @file
+ * @brief       UDP implementation
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @}
+ */
+
+#include <stdint.h>
+#include <errno.h>
+
+#include "kernel.h"
+#include "byteorder.h"
+#include "msg.h"
+#include "thread.h"
+#include "net/ng_udp.h"
+#include "net/ng_netbase.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/**
+ * @brief   Save the UDP's thread PID for later reference
+ */
+static kernel_pid_t _pid = KERNEL_PID_UNDEF;
+
+/**
+ * @brief   Allocate memory for the UDP thread's stack
+ */
+static char _stack[NG_UDP_STACK_SIZE];
+
+/**
+ * @brief   Calculate the UDP checksum dependent on the network protocol
+ *
+ * @note    If the checksum turns out to be 0x0000, the function returns 0xffff
+ *          as specified in RFC768
+ *
+ * @param[in] pkt           pointer to the packet in the packet buffer
+ * @param[in] pseudo_hdr    pointer to the network layer header
+ * @param[in] payload       pointer to the payload
+ *
+ * @return                  the checksum of the pkt
+ * @return                  0 on error
+ */
+uint16_t _calc_csum(ng_pktsnip_t *hdr, ng_pktsnip_t *pseudo_hdr,
+                    ng_pktsnip_t *payload)
+{
+    switch (pseudo_hdr->type) {
+#ifdef MODULE_NG_IPV6
+        case NG_NETTYPE_IPV6:
+            return _calc_csum_ipv6(hdr, pseudo_hdr, payload);
+#endif
+        default:
+            return -ENOENT;
+    }
+}
+
+uint16_t _calc_csum_ipv6(ng_pktsnip_t *udp, ng_pktsnip_t *ip,
+                         ng_pktsnip_t *payload)
+{
+    uint16_t csum = 0;
+
+    /* process the payload */
+    for (int i = 0; i < (payload->size / 2); i++) {
+        csum ^= ((uint16_t *)(payload->data))[i];
+    }
+    if (payload->size & 1) {
+        csum ^= ((uint16_t)((uint8_t *)(payload->data))[payload->size - 1]) << 8;
+    }
+    /* process applicable UDP header bytes */
+    csum ^= ((uint16_t *)(udp->data))[0];
+    csum ^= ((uint16_t *)(udp->data))[1];
+    csum ^= ((uint16_t *)(udp->data))[2];
+    /* process ipv6 pseudo header */
+    for (int i = 0; i < 20; i++) {
+        csum ^= ((uint16_t *)(ip->data))[i];
+    }
+    /* invert checksum if 0 */
+    if (csum == 0) {
+        return 0xffff;
+    }
+    return csum;
+}
+
+void _receive(ng_pktsnip_t *pkt)
+{
+    ng_pktsnip_t *udp;
+    ng_udp_hdr_t *hdr;
+    uint16_t csum;
+    uint32_t port;
+    ng_netreg_entry_t *sendto;
+
+    /* mark UDP header */
+    pkt = ng_pktbuf_start_write(pkt);
+    if (pkt == NULL) {
+        DEBUG("udp: unable to get write access to packet\n");
+        return;
```

Comment:
wait, never mind. Just adding `ng_pktbuf_release(pkt)` does indeed not make sense, but changing the name of the returned pointer  for `ng_pktbuf_start_write()` and then checking that for being `NULL` does make sense. Will fix it.